### PR TITLE
Adds a chart for the external-dns service

### PR DESCRIPTION
**Pull request by rk295**
Hi,

This is my first chart, so please go easy. I've tried to cover all the options provided by `external-dns` with the exception of the `file` ones because I couldn't figure out how to make that work nicely (suggestions welcome). 

I only have AWS & R53 in which to test, but it is working fine for me, will obviously require more testing though.

### CODE REVIEW

**Comment by sstarcher** [comment]
@rk295 I believe this chart meets the requirements to be in the stable folder.
https://github.com/kubernetes/charts/blob/master/CONTRIBUTING.md#technical-requirements

**Comment by sstarcher** [comment]
Could you add support for podAnnotations.  This would enable people to use kube2iam to supply the permissions to the pod.

https://github.com/kubernetes/charts/blob/bcb4436d85a30f9912fd59dde502407fe2e36f23/stable/spartakus/templates/deployment.yaml#L14

* Deployment
```
spec:
  replicas: {{ .Values.replicaCount }}
  template:
    metadata:
    {{- if .Values.podAnnotations }}
      annotations:
{{ toYaml .Values.podAnnotations | indent 8}}
```

* Values
```
## Annotations to be added to pods
##
podAnnotations: {}
```

**Comment by sstarcher** [comment]
After taking a closer look it should this will also need a ServiceAccount, ClusterRoleBindings as it is using the kubernetes api and RBAC is enabled in 1.6 by default.

I don't know what specific permissions this will need though.
https://github.com/kubernetes/charts/blob/2f5852105ed27f1c2ffb701b2482a83834697d7b/incubator/istio/templates/rbac.yaml#L3

**Comment by sstarcher** [comment]
As for the file, you can support the azureConfigFile via a ConfigMap+Volume Mount

Something like the following

* ConfigMap
```
{{ if .Values.azureConfig -}}
apiVersion: v1
kind: ConfigMap
data: |
{{ toJson .Values.azureConfig | indent 2 }}
{{- end }}
```
https://github.com/kubernetes/charts/blob/224b2f7a873173b51212b8f36cd4130d3b789adf/stable/grafana/templates/configmap.yaml#L13

* Values
```
# azureConfig: ""
```

* Deployment
```
    spec:
      containers:
        - name: blah
{{ if .Values.azureConfig -}}
          volumeMounts:
            - name: config-volume
              mountPath: "/etc/azure"
      volumes:
        - name: config-volume
          configMap:
            name: {{ template "fullname" . }}-config
{{- end }}
```

File: incubator/external-dns/Chart.yaml
Comment (linki) [inline incubator/external-dns/Chart.yaml]: We can enrich the yaml similar to https://github.com/kubernetes/charts/blob/master/stable/mediawiki/Chart.yaml. This allows us to have both the chart version and external-dns's version in here.

**Comment by rk295** [comment]
@linki & @sstarcher sure thing, I think I can address most of the changes above. The one I'm a little out of my depth about is the RBAC one, I have no knowledge of RBAC in K8s and have no cluster with it enabled to test. So I think for now I'll be able to address the other comments but not that one. 

Up to you how to play it from here, shall I go ahead and fix the others or shall I close the PR for now until somebody can write the RBAC related templates?

File: incubator/external-dns/values.yaml
Comment (linki) [inline incubator/external-dns/values.yaml]: This could maybe be done via
```yaml
  sources:
  - service
  - ingress
```
and then just iterate over it writing out `--source=${i}` lines in the args section of the deployment template?

File: incubator/external-dns/templates/deployment.yaml
Comment (linki) [inline incubator/external-dns/templates/deployment.yaml]: I think `.Values.image.repository` doesn't need to be guarded. If someone has it on dockerhub she could just use `repository: docker.io`?

File: incubator/external-dns/values.yaml
Comment (linki) [inline incubator/external-dns/values.yaml]: default value of ExternalDNS since `v0.3.0` is `--policy=sync`. I think we should use that as a default here as well.

File: incubator/external-dns/templates/deployment.yaml
Comment (linki) [inline incubator/external-dns/templates/deployment.yaml]: I think it is possible to not guard this. We could then uncomment `compatibility:` in values.yaml. This would lead to something like:

```yaml
args:
- --compatibility=
```

in the rendered manifest when no compatibility is used, which does work as expected (no compatibility). This approach works for other flags as well, e.g. `--namespace` (see https://github.com/kubernetes/charts/blob/master/stable/chaoskube/values.yaml#L23).

However, it clutters the manifest a bit but would cleanup the template and values.yaml. Not sure about this, let me know what you think @rk295.

File: incubator/external-dns/templates/deployment.yaml
Comment (linki) [inline incubator/external-dns/templates/deployment.yaml]: iterating over `.Values.sources` and printing `--source={current iterator value}` seems nicer and more extensible to me.

File: incubator/external-dns/values.yaml
Comment (sstarcher) [reply incubator/external-dns/values.yaml]: if the interest is using the default would it be best to comment out this line

**Comment by sstarcher** [comment]
@rk295 sounds good on skipping the RBAC portion.  If you can add the ServiceAccount stuff it would be easy for a user to bind extra permissions to the pod.

It would look like
https://github.com/kubernetes/charts/blob/2f5852105ed27f1c2ffb701b2482a83834697d7b/incubator/istio/templates/ca-deployment.yaml#L25

and
https://github.com/kubernetes/charts/blob/2f5852105ed27f1c2ffb701b2482a83834697d7b/incubator/istio/templates/ca-service-account.yaml

File: incubator/external-dns/templates/deployment.yaml
Comment (sstarcher) [inline incubator/external-dns/templates/deployment.yaml]: this will include the chart: label that will include the chart version.  This will make upgrades fail.

File: incubator/external-dns/templates/deployment.yaml
Comment (sstarcher) [inline incubator/external-dns/templates/deployment.yaml]: should include the ability to apply resource restrictions. https://github.com/kubernetes/charts/blob/master/stable/uchiwa/templates/deployment.yaml#L21

File: incubator/external-dns/templates/deployment.yaml
Comment (rk295) [reply incubator/external-dns/templates/deployment.yaml]: Is there a 'standard' set of labels charts should apply? It makes sense if they are uniform to me.

Comment (sstarcher) [reply incubator/external-dns/templates/deployment.yaml]: generally 4 labels are applied to the deployment and only 3 to the pods.

Comment (sstarcher) [reply incubator/external-dns/templates/deployment.yaml]: the biggest thing that matters is the label applied to the pod should not change on an upgrade and it will if you include the version.

Comment (rk295) [reply incubator/external-dns/templates/deployment.yaml]: 👍

File: incubator/external-dns/values.yaml
Comment (rk295) [reply incubator/external-dns/values.yaml]: My (slight) preference here is to explicitly set it, that prevents a situation where a user bumps the version of `external-dns` which may change it's default in the future meaning the state the chart leaves the service is would be unknown, i.e. a user would have to go and read the `external-dns` docs to discover the default for *their* chosen version.  I'm not hugely tided to that view though :)

File: incubator/external-dns/templates/deployment.yaml
Comment (rk295) [reply incubator/external-dns/templates/deployment.yaml]: My (again, slight) preference is to have the manifest as clean as possible, i.e. only those options which the user explicitly set.  Happy to do as you suggest though, if you think that is better.

**Comment by rk295** [comment]
@sstarcher & @linki - I've addressed a bunch of the comments above, a few still to sort though.

File: incubator/external-dns/values.yaml
Comment (rk295) [reply incubator/external-dns/values.yaml]: D'oh, I've misspelt it, sink. It sync. It's late! Will fix up tomorrow.

**Comment by SamClinckspoor** [comment]
If you want to include RBAC, we use the following RBAC setup for external-dns. We don't have a lot of runtime with it yet but so far it seems to do the job:
```yaml
---
apiVersion: rbac.authorization.k8s.io/v1beta1
kind: ClusterRoleBinding
metadata:
  name: external-dns
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: ClusterRole
  name: external-dns
subjects:
  - kind: ServiceAccount
    name: external-dns
    namespace: default
---
apiVersion: v1
kind: ServiceAccount
metadata:
  name: external-dns
---
apiVersion: rbac.authorization.k8s.io/v1beta1
kind: ClusterRole
metadata:
  name: external-dns
rules:
- apiGroups:
  - ""
  - "extensions"
  resources:
  - services
  - ingresses
  verbs:
  - list
```

File: incubator/external-dns/templates/deployment.yaml
Comment (linki) [reply incubator/external-dns/templates/deployment.yaml]: Not sure either. It's your call.

File: stable/external-dns/README.md
Comment (linki) [inline stable/external-dns/README.md]: :satisfied:

File: stable/external-dns/values.yaml
Comment (linki) [inline stable/external-dns/values.yaml]: reminder: change to `sync`

File: stable/external-dns/values.yaml
Comment (linki) [inline stable/external-dns/values.yaml]: Since we're using helm could we have a smarter default value here? maybe the release name?

File: stable/external-dns/values.yaml
Comment (linki) [inline stable/external-dns/values.yaml]: > (default: default)

seems odd

File: stable/external-dns/values.yaml
Comment (linki) [inline stable/external-dns/values.yaml]: I think `1` wouldn't be valid. It has to be something like `1m`.

File: stable/external-dns/values.yaml
Comment (linki) [inline stable/external-dns/values.yaml]: defining `debug: false` and at the same time commenting out `dryRun` doesn't seem right. Let's pick either approach.

File: stable/external-dns/values.yaml
Comment (linki) [inline stable/external-dns/values.yaml]: Setting `once: true` is very unlikely to be used by a helm user. Maybe we can just leave it out.

File: stable/external-dns/values.yaml
Comment (rk295) [reply stable/external-dns/values.yaml]: @linki yep, I like that plan. I'll do that.

Comment (rk295) [reply stable/external-dns/values.yaml]: QQ about that, how do I reference `{{ .Release.Name }}` in `values.yaml`? Or should I default it in the actual template?

File: stable/external-dns/values.yaml
Comment (rk295) [reply stable/external-dns/values.yaml]: Correct, I've fixed.

File: stable/external-dns/values.yaml
Comment (rk295) [reply stable/external-dns/values.yaml]: Agreed, I've comment both on the basis they then won't end up in the manifest, but the user can set them if desired.

File: stable/external-dns/values.yaml
Comment (rk295) [reply stable/external-dns/values.yaml]: Agreed, I just blindly added support for all the options without thinking that one through :)

File: stable/external-dns/README.md
Comment (sstarcher) [inline stable/external-dns/README.md]: This is actually a list in the values.yaml
```
source:
 - fake
 - ingress
```

File: stable/external-dns/values.yaml
Comment (linki) [reply stable/external-dns/values.yaml]: Oh yes, I think you have to do it in the template and leave it blank here.

File: stable/external-dns/README.md
Comment (jrnt30) [inline stable/external-dns/README.md]: Currently this PR is into `stable/`

File: stable/external-dns/templates/_helpers.tpl
Comment (jrnt30) [inline stable/external-dns/templates/_helpers.tpl]: I don't think there has been great consolidation on this, however when you have template helper methods and you include a chart as a dependency, these names actually get added to the same global space and collide. In the absence of a good "common chart" I would suggest prefixing these helper methods with the chart name or something else similarly unique so they don't have the potential to collide with one another. 

Ex. 
```
{{- define "external-dns.fullname" -}}
{{- define "external-dns.labels" -}}
```

**Review by jrnt30 — CHANGES_REQUESTED** [review]

**Comment by ghost** [comment]
Any blockers left?

**Comment by taylorsmithgg** [comment]
Just tested the PR, this is the only output:
```
time="2017-07-17T03:52:47Z" level=fatal msg="flag parsing error: enum value must be one of sync,upsert-only, got ''"
```

**Comment by rk295** [comment]
@taylorsmithgg my bad, since the PR work has had me on other things. I'll get back to - and tidy this up - in the next day or so.

File: stable/external-dns/Chart.yaml
Comment (linki) [inline stable/external-dns/Chart.yaml]: Let's use a better description here. Maybe something like the one on the repo itself:

> Configure external DNS servers (AWS Route53, Google CloudDNS and others) for Kubernetes Ingresses and Services

File: stable/external-dns/Chart.yaml
Comment (linki) [inline stable/external-dns/Chart.yaml]: Let's go with version `0.4`. As the final version isn't released yet you can use `v0.4.0-alpha.1` for that matter. I'll make sure it gets released before the chart is done.

We changed some flags in `v0.4` and then we don't have to change them again.

Comment (rk295) [reply stable/external-dns/Chart.yaml]: 👍 I've gone for v0.4.0-alpha.2 if thats ok? I couldn't find any flags that had changed, do you have a link to any docs about that?

p.s. I'll push my changes shortly.

File: stable/external-dns/Chart.yaml
Comment (rk295) [reply stable/external-dns/Chart.yaml]: 👍

**Comment by rk295** [comment]
@taylorsmithgg just looked into this and policy has to be either `upsert-only` or `sync`, I had it commented out in `values.yaml` so an empty `--policy=` option was being passed. I'm now defaulting to `upsert-only`, allowing the user of the chart to override if required.  I'll push that change shortly.

File: stable/external-dns/README.md
Comment (rk295) [reply stable/external-dns/README.md]: 👍

File: stable/external-dns/templates/_helpers.tpl
Comment (rk295) [reply stable/external-dns/templates/_helpers.tpl]: I missed this, now added.

**Comment by rk295** [comment]
@sstarcher I believe I've addressed the service account with the most recent commit, could you check and confirm please?

**Comment by sstarcher** [comment]
@rk295 excellent, I'll review it tomorrow thanks for all of the hard work.

File: stable/external-dns/values.yaml
Comment (sstarcher) [inline stable/external-dns/values.yaml]: I would recommend ```serviceAccountEnabled: false```

I originally assumed this was the name of the service account.

**Comment by sstarcher** [comment]
LGTM - lets get this in and add to it as needed

File: stable/external-dns/values.yaml
Comment (rk295) [reply stable/external-dns/values.yaml]: kk, just pushed a commit with that change.

File: stable/external-dns/templates/deployment.yaml
Comment (mgoodness) [inline stable/external-dns/templates/deployment.yaml]: Does each of the following arguments need to have its own value? I'm generally in favor of providing only the arguments that are commonly required to run the application, then providing an `extraArgs` map for users to provide additional ones. See [here](https://github.com/kubernetes/charts/blob/master/stable/prometheus/templates/server-deployment.yaml#L51) and [here](https://github.com/kubernetes/charts/blob/master/stable/kube2iam/templates/daemonset.yaml#L36) as examples.

My concern is that the list of options will continue growing, regardless of how commonly-used they are, making the chart unwieldy to new users. Better to keep it simple yet extendable.

File: stable/external-dns/templates/deployment.yaml
Comment (mgoodness) [inline stable/external-dns/templates/deployment.yaml]: Is there an advantage to explicitly setting the strategy to `Recreate` rather than not providing one, thus defaulting to `RollingUpdate`?

File: stable/external-dns/templates/deployment.yaml
Comment (mgoodness) [inline stable/external-dns/templates/deployment.yaml]: Delete. I've seen this cause problems.

File: stable/external-dns/templates/deployment.yaml
Comment (mgoodness) [inline stable/external-dns/templates/deployment.yaml]: Merge `image.name` into `image.repository`. We don't need both.

File: stable/external-dns/README.md
Comment (mgoodness) [inline stable/external-dns/README.md]: `consul` -> `external-dns`

File: stable/external-dns/values.yaml
Comment (mgoodness) [inline stable/external-dns/values.yaml]: `Uchiwa` -> `external-dns`

**Review by mgoodness — CHANGES_REQUESTED** [review]
This looks pretty good to me. A few minor changes and a couple questions to address. Then we should be able to get it merged pretty quickly.

**Comment by mgoodness** [comment]
Regarding RBAC, I'll happily add that (and tinker a bit with the ServiceAccount) once this is merged.

File: stable/external-dns/templates/deployment.yaml
Comment (jrnt30) [inline stable/external-dns/templates/deployment.yaml]: Debug is a bool flag, setting it to a value explicitly does not work

File: stable/external-dns/templates/deployment.yaml
Comment (jrnt30) [inline stable/external-dns/templates/deployment.yaml]: Dry run is also a bool flag, setting it to a value explicitly does not work

File: stable/external-dns/values.yaml
Comment (jrnt30) [inline stable/external-dns/values.yaml]: The way the helper works, this doesn't actually override the full service name it just overrides the suffix. That may be fine, but it was slightly confusing upon first use

File: stable/external-dns/Chart.yaml
Comment (jordanjennings) [inline stable/external-dns/Chart.yaml]: Release 0.4.1 is out now: https://github.com/kubernetes-incubator/external-dns/releases

File: stable/external-dns/templates/deployment.yaml
Comment (rk295) [reply stable/external-dns/templates/deployment.yaml]: 👍

File: stable/external-dns/templates/deployment.yaml
Comment (rk295) [reply stable/external-dns/templates/deployment.yaml]: I honestly don't know why I had this set to `Recreate`.  I'd prefer the default too, changed.

File: stable/external-dns/templates/deployment.yaml
Comment (rk295) [reply stable/external-dns/templates/deployment.yaml]: 👍

File: stable/external-dns/README.md
Comment (rk295) [reply stable/external-dns/README.md]: 👍

File: stable/external-dns/values.yaml
Comment (rk295) [reply stable/external-dns/values.yaml]: 👍

File: stable/external-dns/templates/deployment.yaml
Comment (rk295) [reply stable/external-dns/templates/deployment.yaml]: lol, good spot. Me being to hasty I expect!

File: stable/external-dns/templates/deployment.yaml
Comment (rk295) [reply stable/external-dns/templates/deployment.yaml]: As above!

File: stable/external-dns/values.yaml
Comment (rk295) [reply stable/external-dns/values.yaml]: I'm open to suggestions on this one. I pondered a few ways of doing it. I guess if there is some sort of (formal or otherwise) 'standard' used in Charts I'd look to follow that.

File: stable/external-dns/Chart.yaml
Comment (rk295) [reply stable/external-dns/Chart.yaml]: 👍

File: stable/external-dns/templates/deployment.yaml
Comment (rk295) [reply stable/external-dns/templates/deployment.yaml]: Agreed, I prefer this approach. I've addressed all the other outstanding comments and will have a little tinker tonight to see if I can make this work and to pick a sane set of default options.

Comment (rk295) [reply stable/external-dns/templates/deployment.yaml]: Oh! that was far easier than I thought. Thanks for the links to examples :) All comments addressed now.

File: stable/external-dns/templates/deployment.yaml
Comment (linki) [reply stable/external-dns/templates/deployment.yaml]: The reason we run it this way is because ExternalDNS shouldn't run concurrently. When usign the rolling updating this can happen though, whereas `Recreate` makes sure the current pod is fully terminated before the next one spawns (at least this worked fine on Kube 1.7).

However, the impact of a temporary concurrent run shouldn't have any harming effect, so 👍 .

File: stable/external-dns/README.md
Comment (linki) [inline stable/external-dns/README.md]: This is actually `v0.4.2`.

File: stable/external-dns/README.md
Comment (linki) [inline stable/external-dns/README.md]: Full list of supported providers as of `v0.4`: `aws, google, azure, cloudflare, digitalocean, inmemory`

**Comment by linki** [comment]
Just found two minor inconsistencies in the readme. Then LGTM 👍

File: stable/external-dns/README.md
Comment (rk295) [reply stable/external-dns/README.md]: 👍

File: stable/external-dns/README.md
Comment (rk295) [reply stable/external-dns/README.md]: 👍

**Comment by rk295** [comment]
@linki both those addressed now.

File: stable/external-dns/templates/deployment.yaml
Comment (rk295) [reply stable/external-dns/templates/deployment.yaml]: Aha, @linki so that's where I got that from, good to know. I'm happy to flip it back if you think thats better? Your call.

**Review by linki — APPROVED** [review]
lgtm

**Review by mgoodness — APPROVED** [review]

**Comment by mgoodness** [comment]
Made a few changes of my own, mostly just to bring the chart up to spec with our patterns. Biggest changes are that I renamed `source` to `sources` and `domainFilter` to `domainFilters` since both can be lists.

If you don't object to the changes, I'll get this merged right away and work on RBAC & IAM.

BTW, kudos for the `labels` helper function! Very nice way to DRY.

**Comment by rk295** [comment]
@mgoodness 👍 to your changes. Good spot with `domainFilter`. I'm excited to get this in, I can finally remove the fork of it from our private repo.

**Comment by rk295** [comment]
Re your first comment, it does not come up without a key. I'll look at the chart you linked to and try and handle that more gracefully!

**Pull request merged**
